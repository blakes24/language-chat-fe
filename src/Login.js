import React from "react";
import { makeStyles } from "@material-ui/core/styles";
import Container from "@material-ui/core/Container";
import AuthForm from "./AuthForm";
import { useHistory } from "react-router-dom";

const useStyles = makeStyles({
  root: {
    textAlign: "center",
  },
});

function Login() {
  const classes = useStyles();
  const history = useHistory();

  function login(values) {
    console.log(values);
    history.push("/");
  }
  function facebookLogin(user) {
    console.log(user);
    history.push("/");
  }
  function googleLogin(user) {
    console.log(user);
    history.push("/");
  }

  return (
    <Container maxWidth="sm" className={classes.root}>
      <h1>Log In</h1>
      <AuthForm
        text="Log In"
        handleFacebook={facebookLogin}
        handleGoogle={googleLogin}
        handleSubmit={login}
      />
    </Container>
  );
}

export default Login;
