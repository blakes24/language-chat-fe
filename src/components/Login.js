import { useContext, useState } from "react";
import { makeStyles } from "@material-ui/core/styles";
import Container from "@material-ui/core/Container";
import AuthForm from "./AuthForm";
import { useHistory } from "react-router-dom";
import UserContext from "../helpers/UserContext";
import ChatApi from "../helpers/api";
import Card from "@material-ui/core/Card";

const useStyles = makeStyles((theme) => ({
  root: {
    textAlign: "center",
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    height: "90vh",
    minHeight: "500px",
  },
  card: {
    padding: "1rem",
  },
  error: {
    color: theme.palette.error.main,
  },
}));

function Login() {
  const classes = useStyles();
  const history = useHistory();
  const { setToken } = useContext(UserContext);
  const [error, setError] = useState(null);

  function handleFail() {
    setError("Unable to verify account.");
  }

  async function login(values) {
    let res = await ChatApi.getToken(values);
    setToken(res);
    history.push("/");
  }
  async function facebookLogin(user) {
    setError(null);
    try {
      let res = await ChatApi.getToken({
        provider: "facebook",
        token: user._token.idToken,
      });
      setToken(res);
      history.push("/");
    } catch (err) {
      setError("Account not found.");
    }
  }
  async function googleLogin(user) {
    setError(null);
    try {
      let res = await ChatApi.getToken({
        provider: "google",
        token: user._token.idToken,
      });
      setToken(res);
      history.push("/");
    } catch (err) {
      setError("Account not found");
    }
  }

  return (
    <Container maxWidth="sm" className={classes.root}>
      <Card className={classes.card}>
        <h1>Log In</h1>
        <AuthForm
          text="Log In"
          handleFacebook={facebookLogin}
          handleGoogle={googleLogin}
          handleSubmit={login}
          handleFail={handleFail}
        />
        {error && <p className={classes.error}>{error}</p>}
      </Card>
    </Container>
  );
}

export default Login;
