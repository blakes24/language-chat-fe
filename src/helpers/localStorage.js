export function setLocalStorage(key, val) {
  window.localStorage.setItem(key, JSON.stringify(val));
}

export function getFromLocalStorage(key, defaultValue) {
  let value;
  try {
    value = JSON.parse(
      window.localStorage.getItem(key) || JSON.stringify(defaultValue)
    );
  } catch (e) {
    console.error(e);
    value = defaultValue;
  }
  return value;
}
