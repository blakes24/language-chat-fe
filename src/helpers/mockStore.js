import configureStore from "redux-mock-store";

const middlewares = [];
const mockStore = configureStore(middlewares);

const testUser = {
  id: 1,
  name: "Test",
  bio: "stuff about mae",
  email: "test@mail.com",
  imageUrl: "https://cdn.fakercloud.com/avatars/happypeter1983_128.jpg",
  active: true,
  speaks: [
    {
      code: "it",
      language: "Italian",
    },
  ],
  learning: [
    {
      code: "ko",
      language: "Korean",
      level: "2",
    },
  ],
};

const initialState = {
  users: {
    current: testUser,
    items: [],
    token: "",
    loading: "idle",
    error: null,
  },
  rooms: { items: [], current: null, loading: "idle", error: null },
  messages: { items: [], loading: "idle", error: null },
  partners: { items: [], loading: "idle", error: null },
};

export const store = mockStore(initialState);
