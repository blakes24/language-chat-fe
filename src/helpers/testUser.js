import UserContext from "./UserContext";

const testUser = {
  id: 1,
  name: "Test",
  bio: "stuff about mae",
  email: "test@mail.com",
  imageUrl: "https://cdn.fakercloud.com/avatars/happypeter1983_128.jpg",
  active: true,
  speaks: [
    {
      code: "it",
      language: "Italian",
    },
  ],
  learning: [
    {
      code: "ko",
      language: "Korean",
      level: "2",
    },
  ],
};

const UserProvider = ({
  children,
  user = testUser,
  setToken = () => console.log("set token"),
}) => <UserContext.Provider value={{ user }}>{children}</UserContext.Provider>;

export { UserProvider };
