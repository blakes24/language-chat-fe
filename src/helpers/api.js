import axios from "axios";
import { BASE_URL } from "../config";
import { getFromLocalStorage } from "./localStorage";

/** API Class: methods used to connect to the API. */

let token = getFromLocalStorage("token", "");

class ChatApi {
  static token = token;

  static async request(endpoint, data = {}, method = "get") {
    console.debug("API Call:", endpoint, data, method);
    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${ChatApi.token}` };
    const params = method === "get" ? data : {};
    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      let message = err.response
        ? err.response.data.error.message
        : "Server error: try again later";
      throw Array.isArray(message) ? message : [message];
    }
  }

  /** Create a new user and get auth token. */

  static async register(userData) {
    let res = await this.request(`auth/register`, userData, "post");
    this.token = res.token;
    return res.token;
  }

  /** Get user details. */

  static async getUser(userId) {
    let res = await this.request(`users/${userId}`);
    return res.user;
  }

  /** Verifies that the social token provided is valid */

  static async verify(token) {
    let res = await this.request(`auth/validate`, token, "post");
    return res;
  }

  /** Get gets token to authenticate future requests to api */

  static async getToken(data) {
    let res = data.email
      ? await this.request(`auth/token`, data, "post")
      : await this.request(`auth/social-token`, data, "post");
    this.token = res.token;
    return res.token;
  }

  /** Get all users, filter by speaks or learning language */

  static async getAllUsers(filters) {
    let res = await this.request(`users`, filters);
    return res.users;
  }

  /** Get a users chat rooms, filter by chat partner */

  static async getRooms({ userId, partnerId }) {
    let queryParams = { partner: partnerId };
    let res = await this.request(`users/${userId}/rooms`, queryParams);
    return res;
  }

  /** Creates a chat room for two users data = {user1, user2} */

  static async createRoom(data) {
    let res = await this.request(`rooms`, data, "post");
    return res;
  }

  /** Gets messages, filter by room */

  static async getMessages(roomId) {
    let queryParams = { room: roomId };
    let res = await this.request(`messages`, queryParams);
    return res.messages;
  }

  /** Update user info. */

  static async updateUser(id, data) {
    let res = await this.request(`users/${id}`, data, "patch");
    return res.user;
  }

  /** Update language(speaks). */

  static async updateSpeaks(id, data) {
    let res = await this.request(`languages/speaks/${id}`, data, "patch");
    return res.language;
  }

  /** Update language(learning). */

  static async updateLearning(id, data) {
    let res = await this.request(`languages/learning/${id}`, data, "patch");
    return res.language;
  }

  /** Get a users partners */

  static async getPartners(userId) {
    let res = await this.request(`users/${userId}/partners`);
    return res.partners;
  }

  /** Adds a partner */

  static async addPartner(userId, data) {
    let res = await this.request(`users/${userId}/partners`, data, "post");
    return res;
  }

  /** Deletes a partner */

  static async deletePartner({ userId, partnerId }, data) {
    let res = await this.request(
      `users/${userId}/partners/${partnerId}`,
      data,
      "delete"
    );
    return res;
  }

  /** Deletes a user */

  static async deleteUser(userId, data) {
    let res = await this.request(`users/${userId}`, data, "delete");
    return res;
  }
}

export default ChatApi;
