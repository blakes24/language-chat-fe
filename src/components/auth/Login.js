import { useState } from "react";
import { makeStyles } from "@material-ui/core/styles";
import Container from "@material-ui/core/Container";
import AuthForm from "./AuthForm";
import { useHistory } from "react-router-dom";
import { useDispatch } from "react-redux";
import { setToken } from "../../store/usersSlice";
import ChatApi from "../../helpers/api";
import Card from "@material-ui/core/Card";
import { setLocalStorage } from "../../helpers/localStorage";

const useStyles = makeStyles((theme) => ({
  root: {
    textAlign: "center",
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    height: "90vh",
    minHeight: "500px",
  },
  card: {
    padding: "1rem",
  },
  error: {
    color: theme.palette.error.main,
  },
}));

function Login() {
  const classes = useStyles();
  const history = useHistory();
  const dispatch = useDispatch();
  const [error, setError] = useState(null);

  function handleFail() {
    setError("Unable to verify account.");
  }

  async function login(values) {
    let res = await ChatApi.getToken(values);
    dispatch(setToken(res));
    setLocalStorage("token", res);
    history.push("/");
  }
  async function facebookLogin(user) {
    setError(null);
    try {
      let res = await ChatApi.getToken({
        provider: "facebook",
        token: user._token.accessToken,
      });
      dispatch(setToken(res));
      setLocalStorage("token", res);
      history.push("/");
    } catch (err) {
      setError("Account not found.");
    }
  }
  async function googleLogin(user) {
    setError(null);
    try {
      let res = await ChatApi.getToken({
        provider: "google",
        token: user._token.idToken,
      });
      dispatch(setToken(res));
      setLocalStorage("token", res);
      history.push("/");
    } catch (err) {
      setError("Account not found");
    }
  }

  return (
    <Container maxWidth="sm" className={classes.root}>
      <Card className={classes.card}>
        <h1>Log In</h1>
        <AuthForm
          text="Log In"
          handleFacebook={facebookLogin}
          handleGoogle={googleLogin}
          handleSubmit={login}
          handleFail={handleFail}
        />
        {error && <p className={classes.error}>{error}</p>}
      </Card>
    </Container>
  );
}

export default Login;
