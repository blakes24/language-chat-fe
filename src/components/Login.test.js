import { render } from "@testing-library/react";
import Login from "./Login";
import { UserProvider } from "../helpers/testUser";
import { MemoryRouter } from "react-router";

test("renders without crashing", () => {
  render(
    <MemoryRouter>
        <UserProvider>
          <Login />
        </UserProvider>
    </MemoryRouter>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <Login />
      </UserProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});