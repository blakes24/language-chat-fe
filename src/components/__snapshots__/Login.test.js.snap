// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root makeStyles-root-14 MuiContainer-maxWidthSm"
  >
    <div
      class="MuiPaper-root MuiCard-root makeStyles-card-15 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <h1>
        Log In
      </h1>
      <form
        autocomplete="off"
        class="makeStyles-root-17"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          style="margin: 8px;"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
            data-shrink="false"
            for="email"
            id="email-label"
          >
            Email
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
              id="email"
              name="email"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-23 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-25"
              >
                <span>
                  Email
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          style="margin: 8px;"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
            data-shrink="false"
            for="password"
            id="password-label"
          >
            Password
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
          >
            <input
              aria-invalid="false"
              autocomplete="current-password"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
              id="password"
              name="password"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-23 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-25"
              >
                <span>
                  Password
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-containedSizeLarge MuiButton-sizeLarge MuiButton-fullWidth"
          style="margin: 8px;"
          tabindex="0"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Log In
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </form>
      <h4
        class="makeStyles-divider-19"
      >
        OR
      </h4>
      <div
        class="makeStyles-root-17"
      />
    </div>
  </div>
</DocumentFragment>
`;
